[SYCL] Add libsycl, a SYCL RT library implementation project (#144372)
llvm/llvm-project: libsycl/src/CMakeLists.txt

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../../runtimes/cmake/Modules")
include(WarningFlags)

function(add_sycl_rt_library LIB_NAME LIB_OBJ_NAME)
  if (NOT LLVM_ENABLE_PIC)
    message( FATAL_ERROR "Position-Independent Code generation is required for libsycl shared library" )
  endif()
  # Add an optional argument so we can get the library name to
  # link with for Windows Debug version
  cmake_parse_arguments(ARG "" "IMPLIB_NAME" "COMPILE_OPTIONS;SOURCES" ${ARGN})

  add_library(${LIB_OBJ_NAME} OBJECT ${ARG_SOURCES})

  # Common compilation step setup

>   target_compile_definitions(${LIB_OBJ_NAME} PRIVATE $<$<BOOL:${MSVC}>:_LIBSYCL_BUILD_SYCL_DLL>)

Reviewer: The dependency on `MSVC` doesn't seem right here. I think `_LIBSYCL_BUILD_SYCL_DLL` should be defined for any compiler targeting Windows given that the check for this macro in `libsycl/include/sycl/__detail/config.hpp` is target (not compiler) specific.

Author: updated in https://github.com/llvm/llvm-project/pull/144372/commits/6e473009602cdf0ce5cb2fb2f4fe27acc60125d2#diff-012c1c75642ff105a8083560c92d539c11a7d62cbe4946820528c77ba8875270

  cxx_add_warning_flags(${LIB_OBJ_NAME} ${LIBSYCL_ENABLE_WERROR} ${LIBSYCL_ENABLE_PEDANTIC})

  target_include_directories(
    ${LIB_OBJ_NAME}
    PRIVATE
      ${CMAKE_CURRENT_SOURCE_DIR}
      ${LIBSYCL_BUILD_INCLUDE_DIR}
  )

  add_library(${LIB_NAME} SHARED
              $<TARGET_OBJECTS:${LIB_OBJ_NAME}>)

  add_dependencies(${LIB_OBJ_NAME}
    sycl-headers
  )

  set_target_properties(${LIB_NAME} PROPERTIES LINKER_LANGUAGE CXX)

  if (WIN32)

>     target_link_libraries(${LIB_NAME} PRIVATE shlwapi)

Reviewer: The dependency on `shlwapi` probably isn't needed at this time.
```suggestion
```

Reviewer: There is a chance that the dependency won't be needed once the code that currently requires it is put up for review. LLVM already provides abstractions for many of the features provided by shlwapi. Depending on what functionality is needed, it might make more sense to add an abstraction for it in one of the low-level LLVM libraries like `llvm/Support`.

Author: removed in https://github.com/llvm/llvm-project/pull/144372/commits/6e473009602cdf0ce5cb2fb2f4fe27acc60125d2#diff-012c1c75642ff105a8083560c92d539c11a7d62cbe4946820528c77ba8875270

    if (ARG_IMPLIB_NAME)
      add_custom_command(
        TARGET ${LIB_NAME} POST_BUILD
        COMMAND ${CMAKE_COMMAND} -E copy
        ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/${LIB_NAME}.lib ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/${ARG_IMPLIB_NAME}.lib
        COMMENT "Creating version-agnostic copy of the import library.")
      install(
        FILES ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/${ARG_IMPLIB_NAME}.lib
        DESTINATION "lib${LLVM_LIBDIR_SUFFIX}" COMPONENT libsycl)
    endif()
  endif()

>   if (MSVC)

Reviewer: The `else` block suggests this should be checking `WIN32`, not `MSVC` specifically.
```suggestion
  if (WIN32)
```

Author: I will update all similar places. These difference in intel/llvm came from https://github.com/intel/llvm/pull/7859, a specific build environment. I do agree that these details may be not really needed at this point.

      install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${LIB_NAME}.stripped.pdb"
              DESTINATION ${CMAKE_INSTALL_PREFIX}/bin
              RENAME "${LIB_NAME}.pdb"
              COMPONENT ${LIB_NAME}
              OPTIONAL)
    endif()
  else()
    target_compile_options(${LIB_OBJ_NAME} PUBLIC
                           -fvisibility=hidden -fvisibility-inlines-hidden)

    if (UNIX AND NOT APPLE)
      set(linker_script "${CMAKE_CURRENT_SOURCE_DIR}/ld-version-script.txt")
      target_link_libraries(
          ${LIB_NAME} PRIVATE "-Wl,--version-script=${linker_script}")
      set_target_properties(${LIB_NAME} PROPERTIES LINK_DEPENDS ${linker_script})
    endif()
  endif()

  find_package(Threads REQUIRED)

  target_link_libraries(${LIB_NAME}
    PRIVATE
      ${CMAKE_DL_LIBS}
      ${CMAKE_THREAD_LIBS_INIT}
  )

  set_target_properties(${LIB_NAME} PROPERTIES
                        VERSION ${LIBSYCL_VERSION_STRING}
                        SOVERSION ${LIBSYCL_VERSION_STRING})
endfunction(add_sycl_rt_library)

set(LIBSYCL_SOURCES
    "platform.cpp"
)

if (WIN32)
set(LIB_NAME "sycl${LIBSYCL_MAJOR_VERSION}")
else()
set(LIB_NAME "sycl")

> endif()

Reviewer: I don't think it makes sense for the DLL name to include version information by default. I don't think other LLVM projects embed version information in their respective DLL names for windows/MSVC builds. I can appreciate that a versioned name might be desired as part of an installable run-time component, but here we're just building as part of the monorepo, so the version information isn't helpful and the inconsistent naming doesn't seem desirable.

I think it should be possible for someone building LLVM to override the default library name though. Perhaps this should be a cached setting.

# Version-agnostic name of the import library, has effect on Windows only.

> set(IMPLIB_NAME "sycl")

Reviewer: I know it is customary for import libraries to drop a leading "lib" prefix when the corresponding library is named, e.g., "libfoo". In this case, the name of the project is "libsycl", so perhaps it should be preserved here?
```suggestion
set(IMPLIB_NAME "libsycl")
```

Reviewer: Use of a different name could help to avoid an unintended use of the DPC++ import library that has the same name.

Author: @tahonermann want to reiterate this discussion. I kept sycl.lib name. 
AFAIU libc++ uses c++.dll + c++.lib for dynamic & import libraries and libc++.lib for static library. That means that if we want to align with them - we don't need any extra "lib" prefix for SYCL RT. we don't have static version of library and don't plan to.

Reviewer: Thanks @KseniyaTikhomirova, that's fine; we can change it later if it becomes an issue. I did notice that libclang is built as `libclang.dll` with `libclang.lib` as its import library; it seems there is already some inconsistency. I'm a little concerned about claiming the `sycl.dll` name for ourselves, but I guess it is no worse than taking `libsycl.so` on Linux/UNIX systems.

add_sycl_rt_library(${LIB_NAME} sycl_object
                    SOURCES ${LIBSYCL_SOURCES} IMPLIB_NAME ${IMPLIB_NAME})

if (WIN32)
  add_library(sycl ALIAS ${LIB_NAME})

>   set_target_properties(${LIB_NAME} PROPERTIES DEBUG_POSTFIX d)

Author: this part is related to the same problem/feature with CRT & release vs debug build on Windows. Will remove for now.

>   set_target_properties(${LIB_NAME} PROPERTIES DEBUG_POSTFIX d)

Author: Although we still need to add "d" postfix, I will align impl with libcxx https://github.com/llvm/llvm-project/blob/main/libcxx/CMakeLists.txt#L685

>   set_target_properties(${LIB_NAME} PROPERTIES DEBUG_POSTFIX d)

Author: updated conditions to set "d" suffix properly. As reference I used https://github.com/intel/llvm/pull/18200/files#diff-8b7c97dbec3e02fda06bd9374dde5ff0f454f6d76b6a620236a9e84de71f08ea that is on review now.

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../../runtimes/cmake/Modules")
include(WarningFlags)

function(add_sycl_rt_library LIB_NAME LIB_OBJ_NAME)
  # Add an optional argument so we can get the library name to
  # link with for Windows Debug version
  cmake_parse_arguments(ARG "" "IMPLIB_NAME" "COMPILE_OPTIONS;SOURCES" ${ARGN})

  add_library(${LIB_OBJ_NAME} OBJECT ${ARG_SOURCES})

  # Common compilation step setup
  target_compile_definitions(${LIB_OBJ_NAME} PRIVATE $<$<BOOL:${MSVC}>:_LIBSYCL_BUILD_SYCL_DLL>)

  target_include_directories(
    ${LIB_OBJ_NAME}
    PRIVATE
      ${CMAKE_CURRENT_SOURCE_DIR}
      ${LIBSYCL_BUILD_INCLUDE_DIR}
  )

  add_library(${LIB_NAME} SHARED

>               $<TARGET_OBJECTS:${LIB_OBJ_NAME}>)

Author: This split is done for SYCL RT unittests (UT) target that is not included to this PR.
1) we do not provide SYCL RT static lib and don't plan to
2) fPIC option is set by default in LLVM (see LLVM_ENABLE_PIC)
3) the usage is that we have object library sycl_object. It is used for shared library "sycl" creation and for UT executable creation (not included to this PR).

>               $<TARGET_OBJECTS:${LIB_OBJ_NAME}>)

Reviewer: That makes since, thanks @KseniyaTikhomirova!

  if (ARG_COMPILE_OPTIONS)
    target_compile_options(${LIB_OBJ_NAME} PRIVATE ${ARG_COMPILE_OPTIONS})
  endif()

>   cxx_add_warning_flags(${LIB_OBJ_NAME} ${LIBSYCL_ENABLE_WERROR} ${LIBSYCL_ENABLE_PEDANTIC})

Author: moved above https://github.com/llvm/llvm-project/pull/144372/commits/83458d1a3783ac1dc26a8e21dae24f158dadd78a

                           -fvisibility=hidden -fvisibility-inlines-hidden)

    if (UNIX AND NOT APPLE)
      set(linker_script "${CMAKE_CURRENT_SOURCE_DIR}/ld-version-script.txt")
      target_link_libraries(
          ${LIB_NAME} PRIVATE "-Wl,--version-script=${linker_script}")
      set_target_properties(${LIB_NAME} PROPERTIES LINK_DEPENDS ${linker_script})
    endif()
  endif()

  find_package(Threads REQUIRED)

  target_link_libraries(${LIB_NAME}
    PRIVATE
      ${CMAKE_DL_LIBS}
      ${CMAKE_THREAD_LIBS_INIT}
  )

  set_target_properties(${LIB_NAME} PROPERTIES
                        VERSION ${LIBSYCL_VERSION_STRING}
                        SOVERSION ${LIBSYCL_VERSION_STRING})
endfunction(add_sycl_rt_library)

set(LIBSYCL_SOURCES
    "platform.cpp"
)

if (WIN32)
set(LIB_NAME "sycl${LIBSYCL_MAJOR_VERSION}")
else()
set(LIB_NAME "sycl")
endif()

# Version-agnostic name of the import library, has effect on Windows only.
set(IMPLIB_NAME "sycl")

add_sycl_rt_library(${LIB_NAME} sycl_object COMPILE_OPTIONS ${LIBSYCL_EXTRA_OPTS}
                    SOURCES ${LIBSYCL_SOURCES} IMPLIB_NAME ${IMPLIB_NAME})

if (WIN32)
  add_library(sycl ALIAS ${LIB_NAME})
endif()

> if (CMAKE_SYSTEM_NAME STREQUAL Windows)

Reviewer: Why is this checking for `CMAKE_SYSTEM_NAME` rather than just checking if `WIN32` is true?

Author: updated https://github.com/llvm/llvm-project/pull/144372/commits/83458d1a3783ac1dc26a8e21dae24f158dadd78a

              RENAME "${PDB_FILENAME}.pdb"
              COMPONENT ${LIB_TARGET_NAME}
              OPTIONAL)
    endif()

    target_compile_options(${LIB_OBJ_NAME} PUBLIC /EHsc)
  else()
    target_compile_options(${LIB_OBJ_NAME} PUBLIC
                           -fvisibility=hidden -fvisibility-inlines-hidden)

    if (UNIX AND NOT APPLE)
      set(linker_script "${CMAKE_CURRENT_SOURCE_DIR}/ld-version-script.txt")
      target_link_libraries(
          ${LIB_TARGET_NAME} PRIVATE "-Wl,--version-script=${linker_script}")
      set_target_properties(${LIB_TARGET_NAME} PROPERTIES LINK_DEPENDS ${linker_script})
    endif()
  endif()

  find_package(Threads REQUIRED)

  target_link_libraries(${LIB_TARGET_NAME}
    PRIVATE
      ${CMAKE_DL_LIBS}
      ${CMAKE_THREAD_LIBS_INIT}
  )

  set_target_properties(${LIB_TARGET_NAME} PROPERTIES
                        VERSION ${LIBSYCL_VERSION_STRING}
                        SOVERSION ${LIBSYCL_VERSION_STRING})
  set_target_properties(${LIB_TARGET_NAME} PROPERTIES OUTPUT_NAME ${LIB_OUTPUT_NAME})
endfunction(add_sycl_rt_library)

set(LIBSYCL_SOURCES
    "platform.cpp"
)

set(LIB_NAME "sycl")
set(LIB_OUTPUT_NAME "${LIB_NAME}")

> if (CMAKE_SYSTEM_NAME STREQUAL Windows)

Reviewer: I think we always want to link with one of the DLL variants of the MSVC RT library. How about we do what libc++ does and detect attempted use of a non-DLL variant? See https://github.com/llvm/llvm-project/blob/617af3cc50c48ec71889b893dd2658764f19abce/libcxx/CMakeLists.txt#L387-L390.
```suggestion
if (CMAKE_SYSTEM_NAME STREQUAL Windows)
  if (NOT CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "DLL$")
    message(FATAL_ERROR "libsycl requires a DLL version of the MSVC CRT.")
  endif()
```

Author: check is added in https://github.com/llvm/llvm-project/pull/144372/commits/e502190b210d7fc1738a7fec623694792dd3a64c

>   if (CMAKE_BUILD_TYPE STREQUAL Debug OR CMAKE_MSVC_RUNTIME_LIBRARY STREQUAL MultiThreadedDebugDLL)

Reviewer: Do we need to check for "MultiThreadedDebugDLL" specifically? I think the libc++ logic looks appropriate to use here. See https://github.com/llvm/llvm-project/blob/617af3cc50c48ec71889b893dd2658764f19abce/libcxx/CMakeLists.txt#L677-L679.
```suggestion
  if ((NOT CMAKE_MSVC_RUNTIME_LIBRARY AND uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
      OR (CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "Debug"))
```
I know we expect `CMAKE_MSVC_RUNTIME_LIBRARY` to always be set currently, but this logic will still work as expected if we relax that restriction later.

Author: we don't expect it to be always set. We are ok with cmake default value MultiThreaded$<$<CONFIG:Debug>:Debug>DLL and handle explicitly set values.

Author: MultiThreadedDebugDLL is the only option that is supported for sycl in Debug mode. Why should we introduce confusion by setting "CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "Debug""?

Reviewer: Hmm, I didn't realize that `MultiThreadedDebugDLL` was the only supported option when building in debug mode. It seems that should be asserted rather than result in the "d" suffix being silently dropped. Should we also ensure that a MT variant is used? How about this?
```
  if (CMAKE_MSVC_RUNTIME_LIBRARY
      AND (NOT CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "^MultiThreaded.*DLL$"))
    message(FATAL_ERROR "libsycl requires a multi-threaded DLL version of the MSVC CRT.")
  endif()
  if (uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
    if (NOT CMAKE_MSVC_RUNTIME_LIBRARY STREQUAL "MultiThreadedDebugDLL")
      message(FATAL_ERROR "Debug builds of libsycl require a multi-threaded debug DLL version of the MSVC CRT.")
    endif()
    set(LIB_OUTPUT_NAME "${LIB_OUTPUT_NAME}d")
  endif()
```

Author: "The intent is that the "d" suffix is appended if, and only if, the target is linked with a debug version of the MSVC CRT, correct?"
yes, it is how we do it now.
I would prefer to leave full name (MultiThreadedDebugDLL) since it represents better that SYCL RT is always a DLL, there is no other options.

Reviewer: The required use of a DLL is already assured at this point.

My concern is the "MultiThreaded" part and exact matching. Though CMake doesn't currently acknowledge it, there are additional MSVC CRT implementations; see the [MSVC `/Qspectre` option](https://learn.microsoft.com/en-us/cpp/build/reference/qspectre) and the special support for it that exists in MSVS for C++ projects via the ["Spectre Mitigation" property](https://learn.microsoft.com/en-us/cpp/build/reference/c-cpp-prop-page#spectre-mitigation). I think it isn't unreasonable that CMake might expose these additional run-time libraries; particularly for the MSBuild/Visual Studio generator. I don't know if the spectre CRT libraries are provided in debug form, but I would rather err on the side of adding the "d" suffix when "Debug" is present in the name than risk not adding it when libsycl is linked with an unknown CRT library.
